ledger/src/blockstore.rs

```diff
@@ -1498,26 +1505,114 @@ impl Blockstore {
                         .expect("Expect database get to succeed"),
                 }
             })
             .collect()
     }
 
+    fn initialize_transaction_status_index(&self) -> Result<()> {
+        self.transaction_status_index_cf
+            .put(0, &TransactionStatusIndexMeta::default())?;
+        self.transaction_status_index_cf
+            .put(1, &TransactionStatusIndexMeta::default())?;
+        // This dummy status improves compaction performance
+        self.transaction_status_cf.put(
+            (2, Signature::default(), 0),
+            &TransactionStatusMeta::default(),
+        )
+    }
+
+    fn toggle_transaction_status_index(
+        &self,
+        batch: &mut WriteBatch,
+        to_slot: Slot,
+    ) -> Result<Option<u64>> {
+        let index0 = self.transaction_status_index_cf.get(0)?;
+        if index0.is_none() {
+            return Ok(None);
+        }
+        let mut index0 = index0.unwrap();
+        let mut index1 = self.transaction_status_index_cf.get(1)?.unwrap();
+
+        let mut w_active_transaction_status_index =
+            self.active_transaction_status_index.write().unwrap();
+        if !index0.frozen && !index1.frozen {
+            index0.frozen = true;
+            *w_active_transaction_status_index = 1;
+            batch.put::<cf::TransactionStatusIndex>(0, &index0)?;
+            Ok(None)
+        } else {
+            let result = if index0.frozen && to_slot > index0.max_slot {
+                debug!("Pruning transaction index 0 at slot {}", index0.max_slot);
+                Some(0)
+            } else if index1.frozen && to_slot > index1.max_slot {
+                debug!("Pruning transaction index 1 at slot {}", index1.max_slot);
+                Some(1)
+            } else {
+                None
+            };
+
+            if result.is_some() {
+                *w_active_transaction_status_index = if index0.frozen { 0 } else { 1 };
+                if index0.frozen {
+                    index0.max_slot = 0
+                };
+                index0.frozen = !index0.frozen;
+                batch.put::<cf::TransactionStatusIndex>(0, &index0)?;
+                if index1.frozen {
+                    index1.max_slot = 0
+                };
+                index1.frozen = !index1.frozen;
+                batch.put::<cf::TransactionStatusIndex>(1, &index1)?;
+            }
+
+            Ok(result)
+        }
+    }
+
+    fn make_transaction_status_index(
+        &self,
+        index: (Signature, Slot),
+        w_active_transaction_status_index: RwLockWriteGuard<u64>,
+    ) -> Result<(u64, Signature, Slot)> {
+        let (signature, slot) = index;
+        if self.transaction_status_index_cf.get(0)?.is_none() {
+            self.initialize_transaction_status_index()?;
+        }
+        let i = *w_active_transaction_status_index;
+        let mut index_meta = self.transaction_status_index_cf.get(i)?.unwrap();
+        if slot > index_meta.max_slot {
```

Comment:
Let's add an assert here that `!index_meta.frozen`, that should always be true by this point right?